// cosjs.socket/library/socket.js
"use strict";
//agent启动脚本
const library  = require('cosjs.library');

//connector
exports = module.exports = function socket_server(opts){
    var port = opts.port;
    opts['online'] = 0;
    var app = require('socket.io')(port,opts);
    var loader = opts.root ? library('loader')(opts.root) : null;
    app.serveClient(false);
    app.on('connection', function (socket) {
        opts['online'] ++;
        var socket_handle = handle_prototype(app,socket);
        var handle_call_bind = handle_call.bind(socket_handle,loader);
        socket.on('message', handle_call_bind);
        socket.on('error', function (){});
        socket.on('disconnect', function (){ opts['online'] -- ;});
    });

    app.emitter = function(){
        defineHandlePrototype('emitter',defineEmitterPrototype.call(this,arguments[0],opts) );
    }

    app.defineHandlePrototype  = defineHandlePrototype;

    return app;
};



//handle加载器,/test/ , /test/index
function handle_call(loader,name){
    if(!loader) {
        return ;
    }
    var arr = Array.prototype.slice.call(arguments,2);
    var method = loader.parse(name);
    if( typeof method != 'function'  ){
        return false;
    }
    method.apply(this, arr);
};



function handle_prototype(app,socket){
    if (!(this instanceof handle_prototype)) {
        return new handle_prototype(app,socket);
    }
    this.app       = app;
    this.socket    = socket;
};

function defineHandlePrototype(name,value){
    if(typeof name == 'object'){
        for(var k in name){
            defineHandlePrototype(k,name[k]);
        }
    }
    else{
        handle_prototype.prototype[name] = value;
    }
}

function defineEmitterPrototype(opts,setting){
    var loader = opts.root ? library('loader')(opts.root) : null;
    var emitter = library("emitter")(opts);
    emitter.app = this;
    this.defineHandlePrototype('emitter',emitter);
    emitter.redisReceiver.on('message', function (name, message) {
        var arr = library('json').parse(message);
        if(!Array.isArray(arr)){
            return;
        }
        var args = [loader,name].concat(arr);
        handle_call.apply(emitter,args)
    });
    if(setting['refresh']){
        setInterval(function(){
            emitter.emit('connector',setting);
        },setting['refresh']);
    }
}
