// cosjs.library/library/redis/base.js
"use strict"

exports = module.exports = class redis_base{
    constructor(RedisConn,PrefixChar) {
        this._RedisConn  = typeof RedisConn == 'function' ? RedisConn() : RedisConn;
        this._RedisMulti = null;
        this._PrefixChar = Array.isArray(PrefixChar) ? PrefixChar.join('|') : PrefixChar;
        this._PrefixSplit = '|';
    }
    
    connect(read){
        if(read){
            return this._RedisConn;
        }
        else {
            return this._RedisMulti || this._RedisConn;
        }
    }

    exec(callback) {
        callback = callback || this.callback;
        if(!this._RedisMulti){
            return callback('redisErr','use exec but multi empty');
        }
        this._RedisMulti.exec(callback);
    }

    multi(){
        if(!this._RedisMulti){
            this._RedisMulti = this._RedisConn.multi();
        }
        return this._RedisMulti;
    }

    del(key, callback) {
        callback = callback || this.callback;
        var hash  = this.hash(key);
        var redis = this.conn(true);
        redis.DEL(hash, callback);
    }

    expire(id,expire,callback){
        callback = callback || this.callback;
        var hash = this.hash(id);
        var redis = this.conn();
        var time = Date.now() / 1000;
        var cmd = expire >= time ? 'expireat':'expire';
        redis[cmd](hash, expire, callback);
    }

    exists(id,callback){
        var hash = this.hash(id);
        var redis = this.conn(true);
        redis.exists(hash, callback);
    }


    hash(id) {
        var arr = [],key=[];
        for(let k of arguments){
            Array.isArray(k) ? arr=arr.concat(k) : arr.push(k);
        }
        arr.forEach(function(k){
            key.push([this._PrefixChar,k].join(this._PrefixSplit));
        });
        return key;
    }

    callback(err,ret){
        return err ? false : ret;
    }
}
