// cosjs.library/lib/crypto.js
var crypto = require('crypto');

function cosCrypto(secret,garbled){

    if (!(this instanceof cosCrypto)) {
        return new cosCrypto(secret,garbled)
    }

    this.cipherType = 'aes-128-cfb';

    this.encode = function(str){
        return exports.encode(str,secret,garbled,this.cipherType);
    }

    this.decode = function(str){
        return exports.decode(str,secret,garbled,this.cipherType);
    }

    this.password = function(str){
        return exports.password(str,secret);
    }

    this.ObjectID = require('./ObjectID');
}


exports = module.exports = cosCrypto;

exports.md5 = function (str){
    var _encrymd5 = crypto.createHash('md5');
    _encrymd5.update(str.toString(),'utf8');
    return _encrymd5.digest('hex');
}

//签名
exports.sign = function (args,secret,length) {
    var keys = Object.keys(args).sort();
    var arr = [];
    keys.forEach(function(k){
        arr.push(k+'='+args[k]);
    });
    arr.push(secret);
    var str = arr.join('&');
    return exports.md5(str).substr(0,length||16);
}
exports.encode =  function(str,secret,garbled,cipherType){
    var garbledStr = '';
    cipherType = cipherType || 'aes-128-cfb';
    if(garbled){
        garbledStr = exports.md5(Date.now().toString()).substr(0,garbled);
    }
    var newSecret = [secret,garbledStr].join('');
    var cipher  = crypto.createCipher('aes-128-cfb',newSecret);
    var enc  = cipher .update(str, "utf8", "hex");
    enc += cipher .final("hex");
    var ret = garbledStr + enc.toString();
    return ret;
}

exports.decode =  function(str,secret,garbled,cipherType){
    var garbledStr,newStr;
    cipherType = cipherType || 'aes-128-cfb';
    if(garbled){
        newStr = str.substr(garbled);
        garbledStr = str.substr(0,garbled);
    }
    else{
        newStr = str;
        garbledStr = '';
    }
    var newSecret = [secret,garbledStr].join('');
    var decipher = crypto.createDecipher(cipherType,newSecret);
    var decrypted = decipher.update(newStr.toString(), 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    return decrypted;
}

//生成密码
exports.password =  function(str,secret){
    var cipher  = crypto.createCipher('aes-128-cfb',secret);
    var enc  = cipher .update(str, "utf8", "hex");
    enc += cipher .final("hex");
    return exports.md5(enc);
}
