// cosjs.library/lib/emitter.js

var EventEmitter = require('events').EventEmitter
    ,util = require('util')


function RedisEmitter(opts) {
    if (!(this instanceof RedisEmitter)) {
        return new RedisEmitter(opts)
    }
    var self = this;
    this.prefix = opts.prefix || ''
    this.pattern = opts.pattern ? true : false;
    this.redisEmitter  = createClient.call(this,opts,'pub');
    this.redisReceiver = createClient.call(this,opts,'sub');

    if(this.pattern) {
        this.redisReceiver.on('pmessage', function (pattern, name, args) {
            pmessage.call(self, pattern, name, args);
        });
    }
    else{
        this.redisReceiver.on('message', function (name, args) {
            message.call(self, name, args);
        });
    }
}

exports = module.exports = RedisEmitter;

util.inherits(RedisEmitter, EventEmitter);


RedisEmitter.prototype.emit = function (name) {
    if (name === 'newListener' || name === 'error') {
        EventEmitter.prototype.emit.apply(this, arguments)
    } else {
        this.publish.apply(this,arguments);
    }
}

RedisEmitter.prototype.publish = function(name) {
    var args = Array.prototype.slice.call(arguments, 1);
    this.redisEmitter.publish(this.prefix + name, JSON.stringify(args));
}


RedisEmitter.prototype.on = RedisEmitter.prototype.subscribe = RedisEmitter.prototype.addListener = function (name, listener) {
    EventEmitter.prototype.addListener.call(this, name, listener);
    if (name === 'error' || name === 'newListener') {
        return
    }
    var cmd = this.pattern ? 'psubscribe' : 'subscribe';
    this.redisReceiver[cmd](this.prefix + name);
}


RedisEmitter.prototype.unsubscribe = RedisEmitter.prototype.removeListener = function (name) {
    EventEmitter.prototype.removeListener.apply(this, arguments);
    if (this.listeners(name).length == 0) {
        var cmd = this.pattern ? 'punsubscribe' : 'unsubscribe';
        this.redisReceiver[cmd](this.prefix + name);
    }
}

RedisEmitter.prototype.removeAllListener = function (name) {
    EventEmitter.prototype.removeAllListener.apply(this, arguments);
    var cmd = this.pattern ? 'punsubscribe' : 'unsubscribe';
    this.redisReceiver[cmd](this.prefix + name);
}


RedisEmitter.prototype.end = RedisEmitter.prototype.quit = function () {
    if (this.hasOwnProperty('redisEmitter')) this.redisEmitter.end();
    if (this.hasOwnProperty('redisReceiver')) this.redisReceiver.end();
}


function createClient(opts,type) {
    var client,redis = opts.redis || require('redis');
    if (opts[type] instanceof (redis.RedisClient)) {
        client = opts[type];
    } else {
        client = redis.createClient(opts.port, opts.host, opts);
    }
    client.on('error', EventEmitter.prototype.emit.bind(this,'error') );
    return client;
}

function message(name,args) {
    if(!args){
        return;
    }
    args = JSON.parse(args)
    if (this.prefix.length) {
        name = name.substring(this.prefix.length);
    }
    args.unshift(name);
    EventEmitter.prototype.emit.apply(this, args);
}

function pmessage(pattern,name,args) {
    if(!args){
        return;
    }
    args = JSON.parse(args)
    if (this.prefix.length) {
        name = name.substring(this.prefix.length);
        pattern = pattern.substring(this.prefix.length);
    }
    if(pattern.indexOf('*') >=0){
        var arr = [pattern,name].concat(args);
    }
    else{
        var arr = [name].concat(args);
    }
    EventEmitter.prototype.emit.apply(this, arr);
}
