// cosjs.library/lib/loader.js
"use strict";
var fs = require('fs');
//模块加载器


function loadder(path){
    if (!(this instanceof loadder)) {
        return new loadder(path);
    }
    this.ext = ['.js'];
    this.path = [];
    this._safeMode   = false;
    this._moduleCache = {};
    this._handleCache = null;
    if(path) {
        this.path.push(path);
        getFiles.call(this, path, '/');
    }
}

exports = module.exports = loadder;

loadder.prototype.require = function(name){
    if(name[0] !== '/'){
        name = '/' + name;
    }
    if( !this._moduleCache[name] ){
        return false;
    }
    var file = this._moduleCache[name];
    var mods = require(file);
    return mods;
}


loadder.prototype.parse = function(path){
    if( path.length <=1 ){
        return false;
    }
    var i = path.lastIndexOf('/');
    var mod = this.require(path.substr(0,i));
    if(!mod){
        return false;
    }
    var key = path.substr(i+1);
    if(!key){
        return mod;
    }
    else if(typeof mod != 'object'){
        return false;
    }
    return mod[key]||false;
}



loadder.prototype.addPath = function(path) {
    if(this.path.indexOf(path) >=0){
        return;
    }
    this._handleCache = null;
    this.path.push(path);
    getFiles.call(this, path, '/');
}

loadder.prototype.handle = function(){
    if(this._handleCache){
        return this._handleCache;
    }
    this._handleCache = [];
    for(var k in this._moduleCache){
        var mods = require(this._moduleCache[k]);
        if(typeof mods === 'function'){
            this._handleCache.push(k+'/');
        }
        else{
            for(var m in mods){
                if(typeof mods[m] === 'function'){
                    this._handleCache.push(k+'/'+m);
                }
            }
        }
    }
    return this._handleCache;
}

loadder.prototype.module = function(){
    return this._moduleCache;
}


function getFiles(root,dir) {
    var self = this,stats;
    var path = root + dir;
    try {
        stats = fs.statSync(path);
    }
    catch (e){
        stats = null;
    }
    if(!stats){
        return;
    }
    if(!stats.isDirectory()){
        return;
    }
    var files = fs.readdirSync(path);
    if(!files || !files.length){
        return;
    }
    files.forEach(function (name) {
        filesForEach.call(self,root,dir,name);
    })
}


function filesForEach(root,dir,name){
    var self = this,FSPath = require('path');
    var realPath = FSPath.resolve([root,dir,name].join('/'));
    var realName = dir+name;
    var stats = fs.statSync(realPath);
    if (stats.isDirectory()) {
        return getFiles.call(self,root,realName+'/');
    }

    var ext = FSPath.extname(name);
    if(self.ext.indexOf(ext) >=0){
        var api = realName.replace(ext,'');
        if( self._safeMode && self._moduleCache[api] ){
            console.log('file['+api+'] exist');
        }
        else{
            self._moduleCache[api] = realPath;
        }
    }

}
