// cosjs.loader/index.js
"use strict";
var fs = require('fs');
//模块加载器


function loadder(path){
    if (!(this instanceof loadder)) {
        return new loadder(path);
    }
    this.ext = ['.js'];
    this.path = [];
    this._moduleCache = {};
    if(path) {
        this.path.push(path);
        getFiles.call(this, path, '/');
    }
}

exports = module.exports = loadder;

loadder.prototype.require = function(name){
    if(name[0] !== '/'){
        name = '/' + name;
    }
    if( !this._moduleCache[name] ){
        return false;
    }
    var file = this._moduleCache[name];
    return require(file);
}


loadder.prototype.parse = function(path){
    if( path.length <=1 ){
        return false;
    }
    var i = path.lastIndexOf('/');
    var mod = this.require(path.substr(0,i));
    if(!mod){
        return false;
    }
    var key = path.substr(i+1);
    if(!key){
        return mod;
    }
    else if(typeof mod != 'object'){
        return false;
    }
    return mod[key]||false;
}



loadder.prototype.addPath = function(path) {
    if(this.path.indexOf(path) >=0){
        return;
    }
    this.path.push(path);
    getFiles.call(this, path, '/');
}


function getFiles(root,dir) {
    var self = this,FSPath = require('path');
    var path = root + dir;
    var stats = fs.statSync(path);
    if(!stats.isDirectory()){
        return;
    }
    var files = fs.readdirSync(path);
    if(!files || !files.length){
        return;
    }
    files.forEach(function (name) {
        var realPath = FSPath.resolve(path + '/' + name);
        var realName = dir+name;
        var stats = fs.statSync(realPath);
        if (stats.isDirectory()) {
            getFiles.call(self,root,realName+'/');
        }
        else{
            var ext = FSPath.extname(name);
            if(self.ext.indexOf(ext) >=0){
                var api = realName.replace(ext,'');
                self._moduleCache[api] = realPath;
            }
        }
    })
}
